src/restart_counter.rs

```diff
@@ -0,0 +1,104 @@
+use crate::ProbeId;
+use core::fmt;
+
+/// The persistent sequence number value returned when this
+/// probe is not tracking restarts.
+pub const NO_RESTART_TRACKING_SEQUENCE_NUMBER: u16 = 0;
+
+/// A persistent restart sequence counter
+pub trait RestartCounter {
+    /// Get the next persistent sequence number
+    fn next_sequence_id(&mut self, probe_id: ProbeId) -> u16;
+}
+
+/// C function type for retrieving the next persistent sequence number
+#[allow(non_camel_case_types)]
+pub type next_sequence_id_fn = extern "C" fn(u32, *mut core::ffi::c_void) -> u16;
+
+/// A persistent restart sequence counter provider backed by a C implementation
+pub struct CRestartCounterProvider {
+    /// C interface for retrieving the next persistent sequence number
+    pub iface: next_sequence_id_fn,
+    /// User's state provided to the next_sequence_id_fn function call
+    pub state: *mut core::ffi::c_void,
+}
+
+/// A persistent restart sequence counter provider backed by a Rust implementation
+pub struct RustRestartCounterProvider<'a> {
+    /// Rust interface for retrieving the next persistent sequence number
+    pub iface: &'a mut dyn RestartCounter,
+}
+
+/// A persistent restart sequence counter provider
+pub enum RestartCounterProvider<'a> {
+    /// Do no restart handling for this probe.
+    /// Any events logged after a restart will be seen as duplicates.
+    NoRestartTracking,
+    /// A persistent restart sequence counter provider backed by a C implementation
+    C(CRestartCounterProvider),
+    /// A persistent restart sequence counter provider backed by a Rust implementation
+    Rust(RustRestartCounterProvider<'a>),
+}
+
+impl<'a> From<&'a mut dyn RestartCounter> for RestartCounterProvider<'a> {
+    fn from(r: &'a mut dyn RestartCounter) -> Self {
+        RestartCounterProvider::Rust(RustRestartCounterProvider { iface: r })
+    }
+}
+
+impl<'a> RestartCounterProvider<'a> {
+    fn next_sequence_id(&mut self, probe_id: ProbeId) -> u16 {
```

Comment:
Yep, the more I look at this, the more I think that `Option<NonZeroU16>` should be the return type, since it makes it clear whether we have a valid, order-able sequence number or not.  Then, the user of this function (i.e. the fn in history that deals with storing epochs during initialization or overflows) can deal with signaling further upstream this lack in whichever way is appropriate (whether that's stuffing a placeholder 0 into a ProbeEpoch or some approach)

Reply:
Makes sense to me, updated in 3e47a5b3bc3a9d66a04fd118985624f8044689c3.